### PR TITLE
Cassandra-15971 Improve FQL doc page

### PR DESCRIPTION
**Pull request by polandll**
I've rewritten this page to have better organization and flow. Needs review.

### CODE REVIEW

**Comment by Deepak-Vohra** [comment]
Please verify that all sections that need to be in code font are in code font. As example, the following code snippet is not completely in code font.  
```
replay \
--keyspace querylogkeyspace --results /cassandra/fql/logs/results/replay --store-queries /cassandra/fql/logs/queries/replay -- target 3.91.56.164 /tmp/cassandrafullquerylog 
```

**Comment by polandll** [comment]
Thanks for catching that codeblock. I forgot to indent. :-) I spend much more time in Asciidoc than reStructured text. I'm made the fix.

File: doc/source/new/fqllogging.rst [hunk 1]
```diff
@@ -14,174 +14,169 @@
...
+- Can be enabled, disabled, or reset (to delete on-disk data) using the JMX tool, ``nodetool``.
+- Can configure the settings in either the `cassandra.yaml` file or by using ``nodetool``.
+- Introduces new ``fqltool`` that currently can ``Dump`` the binary logs to a readable format. Other options are ``Replay`` and ``Compare``.
 
-Cassandra 4.0 has a binary full query log based on Chronicle Queue that can be controlled using ``nodetool enablefullquerylog``, ``disablefullquerylog``, and ``resetfullquerylog``. The log contains all queries invoked, approximate time they were invoked, any parameters necessary to bind wildcard values, and all query options. A readable version of the log can be dumped or tailed using the new ``bin/fqltool`` utility. The full query log is designed to be safe to use in production and limits utilization of heap memory and disk space with limits you can specify when enabling the log.
+FQL logs all Cassandra Query Language (CQL) requests, both events that modify the data and those that query.
```
Comment (dcapwell) [inline doc/source/new/fqllogging.rst]: > FQL logs all Cassandra Query Language (CQL)

should be

> FQL logs all successful Cassandra Query Language (CQL)

Checking the code, bad queries (ones which don't parse or cause an exception) will not be logged.

Comment (dcapwell) [reply doc/source/new/fqllogging.rst]: see you say this later with ` Failed or timed-out queries are not logged.`, still good to be clear

File: doc/source/new/fqllogging.rst [hunk 2]
```diff
@@ -14,174 +14,169 @@
...
-    # max_archive_retries: 10
+The ``max_log_size`` option sets the maximum size of the rolled files to retain on disk before deleting the oldest file.  The option must be set to a positive value. The default is 17179869184, or 16 GiB.
+For example, to change the default: ``max_log_size: 34359738368 # 32 GiB``
 
-The ``max_queue_weight`` must be > 0. Similarly ``max_log_size`` must be > 0. An example full query logging options is as follows.
+archive_command
+^^^^^^^^^^^^^^^
 
-::
+The ``archive_command`` option sets the user-defined archive script to execute on rolled log files.
```
Comment (dcapwell) [inline doc/source/new/fqllogging.rst:162]: should state that when not defined, files are deleted

Comment (dcapwell) [reply doc/source/new/fqllogging.rst:162]: Default is `""` which then maps to `org.apache.cassandra.utils.binlog.DeletingArchiver`

```
BinLogArchiver archiver = Strings.isNullOrEmpty(archiveCommand) ? new DeletingArchiver(maxLogSize) : new ExternalArchiver(archiveCommand, path, maxArchiveRetries);
```

File: doc/source/new/fqllogging.rst [hunk 3]
```diff
@@ -14,174 +14,169 @@
...
-    # max_log_size: 17179869184 # 16 GiB
-    ## archive command is "/path/to/script.sh %path" where %path is replaced with the file
- being rolled:
-    # archive_command:
-    # max_archive_retries: 10
+The ``max_archive_retries`` option sets the max number of retries of failed archive commands. The default is 10.
+For example: ``max_archive_retries: 10``
 
-The ``full_query_log_dir`` setting is not within the ``full_query_logging_options`` but still is for full query logging.
+FQL can also be configured using ``nodetool` when enabling the feature, and will override any values set in the `cassandra.yaml` file, as discussed in the next section.
```
Comment (dcapwell) [inline doc/source/new/fqllogging.rst]: nodetool has 2 ` upfront, should be 1

File: doc/source/new/fqllogging.rst [hunk 4]
```diff
@@ -239,74 +234,186 @@ The syntax of the nodetool enablefullquerylog command is as follows:
...
+Use ``nodetool resetfullquerylog`` to stop FQL and clear the log files in the configured directory.
+**IMPORTANT:** Using ``nodetool resetfullquerylog`` will delete the log files! Do not use this command unless you need to delete all log files.
+
+fqltool
+-------
+
+The ``fqltool`` command is used to view (dump), replay, or compare logs.
+``fqltool dump`` converts the binary log files into human-readable format; only the log directory must be supplied as a command-line option.
+
+``fqltool replay`` (CASSANDRA-14618) enables replay of logs.
```
Comment (dcapwell) [inline doc/source/new/fqllogging.rst]: should this link to `https://issues.apache.org/jira/browse/CASSANDRA-14618`?  Previous examples did

File: doc/source/new/fqllogging.rst [hunk 5]
```diff
@@ -239,74 +234,186 @@ The syntax of the nodetool enablefullquerylog command is as follows:
...
+
+fqltool
+-------
+
+The ``fqltool`` command is used to view (dump), replay, or compare logs.
+``fqltool dump`` converts the binary log files into human-readable format; only the log directory must be supplied as a command-line option.
+
+``fqltool replay`` (CASSANDRA-14618) enables replay of logs. 
+The command can run from a different machine or cluster for testing, debugging, or performance benchmarking. 
+The command, run on the same node on which the logs are generated can recreate a dropped database object.
```
Comment (dcapwell) [inline doc/source/new/fqllogging.rst]: > The command, run on the same node on which the logs are generated can recreate a dropped database object

I don't know what this means, what are you trying to say?

**Review by dcapwell — COMMENTED** [review]
Overall LGTM, only small comments

File: doc/source/new/fqllogging.rst (hunk 5 above)
Comment (polandll) [reply doc/source/new/fqllogging.rst]: I didn't write that text originally. I've revised to: The command can also be used to recreate a dropped database object.

**Comment by polandll** [comment]
Thanks @dcapwell - I've incorporated your comments!

File: doc/source/new/fqllogging.rst [hunk 6]
```diff
@@ -239,74 +235,186 @@ The syntax of the nodetool enablefullquerylog command is as follows:
...
+The command can run from a different machine or cluster for testing, debugging, or performance benchmarking. 
+The command can also be used to recreate a dropped database object.
+Use ``fqltool replay`` to record and compare different runs of production traffic against different versions/configurations of Cassandra or different clusters.
+Another use is to gather logs from several machines and replay them in “order” by the timestamps recorded.
+
+The syntax of ``fqltool replay`` is:
+
+::
+
+  fqltool replay [--keyspace <keyspace>] [--results <results>]
```
Comment (smiklosovic) [inline doc/source/new/fqllogging.rst]: @polandll you will likely have merge conflicts somewhere along these lines as this was just merged recently.

https://github.com/apache/cassandra/commit/00c22232d96b6e592fb1d1f32a6b649b49eceb34

Comment (polandll) [reply doc/source/new/fqllogging.rst]: Thanks for the heads-up. I resolved the conflict.

**Comment by polandll** [comment]
@dcapwell - I've added the --replay-ddl-statements flag to the replay command. I added a note to indicate that replay is generally done between clusters, but left the example using a single cluster, just to simplify. I hope that this change meets with your approval.

**Comment by smiklosovic** [comment]
@dcapwell it would be really nice if this is in https://github.com/apache/cassandra/pull/721

I am introducing there getfullqerylog command in nodetool and it touches the documentation yet again so it would have to be added there etc.

**Comment by dcapwell** [comment]
working on it now

File: doc/source/new/fqllogging.rst [hunk 7]
```diff
@@ -498,198 +558,31 @@ The full query logs get listed. Truncated output is as follows:
...
-   <path1> [<path2>...<pathN>]
-  Paths containing the full query logs to replay.
-
-As an example of using ``fqltool replay``, drop a keyspace.
-
-::
+ $ fqltool replay \
+ --keyspace querylogkeyspace --results /cassandra/fql/logs/results/replay \
+ --store-queries /cassandra/fql/logs/queries/replay \
+ -- target 3.91.56.164 \
```
Comment (dcapwell) [inline doc/source/new/fqllogging.rst:581]: extra space

File: doc/source/new/fqllogging.rst [hunk 8]
```diff
@@ -498,198 +558,31 @@ The full query logs get listed. Truncated output is as follows:
...
-  list of argument, (useful when arguments might be mistaken for
-  command-line options
-
-   <path1> [<path2>...<pathN>]
-  Paths containing the full query logs to replay.
-
-As an example of using ``fqltool replay``, drop a keyspace.
-
-::
+ $ fqltool replay \
```
Comment (dcapwell) [inline doc/source/new/fqllogging.rst:578]: since we do drop keyspace, will need `--replay-ddl-statements`.

**Review by dcapwell — COMMENTED** [review]
Overall LGTM.  2 small comments but I can fix that on merge.

**Pull request closed without merging**